File: Workshop 3 Solsystemet-20230210/Python/solar_system.py
import time

import matplotlib.pyplot
import matplotlib.pyplot as plt
from Planet import Planet
from Comet import Comet
from Sun import Sun
import numpy as np


class Solar_system:
    def __init__(self):
        self.planets: Planet = list()
        self.comets: Comet = list()
        self.t = -3000
        self.dt = 10
        self.playing = False
        self.paused = False
        self.axis_lim = [-40, 40, -40, 40]
        self.show_text = True

        self.fig, self.ax = plt.subplots()
        self.fig.canvas.mpl_connect('key_press_event', lambda event: self.buttonpush(event))

        plt.axis('equal')
        plt.title('Solsystemet')
        plt.axis(self.axis_lim)
        self.text = plt.text(
            0.8 * self.axis_lim[1], self.axis_lim[3],
            '', verticalalignment='top')

    def add_planets(self, names, colors, planet_data):
        l = len(names)
        self.planets = list()
        # Sort the planets wrt radius
        idx = np.argsort(np.array(planet_data)[:, 0])
        names = [names[i] for i in idx]
        colors = [colors[i] for i in idx]
        planet_data = [planet_data[i] for i in idx]
        for i in range(l):
            planet = Planet(names[i], colors[i], planet_data[i], self.ax)
            self.planets.append(planet)

    def add_comets(self, names, colors, comet_data):
        l = len(names)
        self.comets = list()
        # Sort the comets wrt radius
        idx = np.argsort(np.array(comet_data)[:, 0])
        names = [names[i] for i in idx]
        colors = [colors[i] for i in idx]
        comet_data = [comet_data[i] for i in idx]
        for i in range(l):
            comet = Comet(names[i], colors[i], comet_data[i], self.ax)
            self.comets.append(comet)

    def add_sun(self, radius):
        self.sun = Sun(radius, self.ax)

    def update(self, t):
        self.t = t
        for planet in self.planets:
            planet.update(t)
        for comet in self.comets:
            comet.update(t)
        if hasattr(self, "sun"):
            self.sun.update()

    def draw_system(self):
        for planet in self.planets:
            planet.ball.set_visible(True)
            if self.show_text and \
                    self.axis_lim[0] <= planet.coordinates[0] <= self.axis_lim[1] and \
                    self.axis_lim[2] <= planet.coordinates[1] <= self.axis_lim[3]:
                planet.text.set_visible(True)
            else:
                planet.text.set_visible(False)

        for comet in self.comets:
            comet.ball.set_visible(True)
            if self.show_text and \
                    self.axis_lim[0] <= comet.coordinates[0] <= self.axis_lim[1] and \
                    self.axis_lim[2] <= comet.coordinates[1] <= self.axis_lim[3]:
                comet.text.set_visible(True)
            else:
                comet.text.set_visible(False)

        if hasattr(self, "sun"):
            self.sun.ball.set_visible(True)
            self.sun.text.set_visible(self.show_text)

        self.text.set_text(
            'q: afslut spillet\n'
            'mellemrum: pause/genoptag\n'
            'piletaster: skalér/zoom\n'
            't: tekster til/fra\n'
            'punktum/komma: hastighed op/ned\n'
            f'dato: {self.time_to_date(self.t)}')
        self.text.set_position([0.9 * self.axis_lim[1], 0.9 * self.axis_lim[3]])

        plt.axis(self.axis_lim)

    def scale_distances(self, scale):
        for planet in self.planets:
            planet.a *= scale
        for comet in self.comets:
            comet.a *= scale

    def scale_radii(self, scale):
        for planet in self.planets:
            planet.radius *= scale
        for comet in self.comets:
            comet.radius *= scale
        if hasattr(self, "sun"):
            self.sun.radius *= scale

    @staticmethod
    def time_to_date(t):
        year = int(t)
        fractional_part = t - year
        day = abs(int(365.25 * fractional_part)) + 1
        if year < 1:
            year = abs(year - 1)
            prefix = 'BC'
        else:
            prefix = 'AD'

        if 0 <= day <= 31:
            month = 'Jan'
        elif 31 < day <= 59:
            month = 'Feb'
            day -= 31
        elif 59 < day <= 90:
            month = 'Mar'
            day -= 59
        elif 90 < day <= 120:
            month = 'Apr'
            day -= 90
        elif 120 < day <= 151:
            month = 'May'
            day -= 120
        elif 151 < day <= 181:
            month = 'Jun'
            day -= 151
        elif 181 < day <= 212:
            month = 'Jul'
            day -= 181
        elif 212 < day <= 243:
            month = 'Aug'
            day -= 212
        elif 243 < day <= 273:
            month = 'Sep'
            day -= 243
        elif 273 < day <= 304:
            month = 'Oct'
            day -= 273
        elif 304 < day <= 334:
            month = 'Nov'
            day -= 304
        else:
            month = 'Dec'
            day -= 334
        if day > 31: day = 31

        date = f'{day}. {month} {year} {prefix}'
        return date

    def run(self, t, dt, refresh_rate):
        self.playing = True
        self.dt = dt
        start_time = time.time()
        while self.playing and t <= 3000:
            plt.pause(refresh_rate)
            if self.paused:
                self.draw_system()
                continue
            self.update(t)
            self.draw_system()
            time_elapsed = (time.time() - start_time) * refresh_rate
            t += self.dt * time_elapsed
            self.fig.canvas.draw()

    def buttonpush(self, event):
        if event.key == 'q':
            self.playing = False
        elif event.key == ' ':
            self.paused = not self.paused
        elif event.key == 'left':
            self.scale_radii(0.5)
        elif event.key == 'right':
            self.scale_radii(2)
        elif event.key == 'down':
            self.axis_lim = [x * 2 for x in self.axis_lim]
        elif event.key == 'up':
            self.axis_lim = [x / 2 for x in self.axis_lim]
        elif event.key == ',':
            self.dt /= 2
            for planet in self.planets:
                planet.trace_length += 20
        elif event.key == '.':
            self.dt *= 2
            for planet in self.planets:
                planet.trace_length -= 20
        elif event.key == 't':
            self.show_text = not self.show_text
            self.show_hide_text()

    def show_hide_text(self):

        self.sun.text.set_visible(self.show_text)

        for planet in self.planets:
            if self.show_text and \
                    self.axis_lim[0] <= planet.coordinates[0] <= self.axis_lim[1] and \
                    self.axis_lim[2] <= planet.coordinates[1] <= self.axis_lim[3]:
                planet.text.set_visible(True)
            else:
                planet.text.set_visible(False)

        for comet in self.comets:
            if self.show_text and \
                    self.axis_lim[0] <= comet.coordinates[0] <= self.axis_lim[1] and \
                    self.axis_lim[2] <= comet.coordinates[1] <= self.axis_lim[3]:
                comet.text.set_visible(True)
            else:
                comet.text.set_visible(False)

        self.text.set_text(['q: afslut spillet',
                            'mellemrum: pause/genoptag',
                            'piletaster: skalér/zoom',
                            't: tekster til/fra',
                            'punktum/komma: hastighed op/ned',
                            ['dato: ', self.time_to_date(self.t)]])
        self.text.set_position([0.9 * self.axis_lim[1], 0.9 * self.axis_lim[3]])


if __name__ == "__main__":
    # # Delopgave 3
    # I denne opgave skal Newtons metode implementeres i filen newtons_method.m
    # Når det er gjort vil den anden model af solsystemet, der opfylder Keplers
    # anden lov, virke. Vi starter med at rydde gemte variable og eksisterende
    # figurer.

    # Planet (og Pluto) data
    planet_data = np.loadtxt('planet_data1.txt', delimiter=',')
    planet_names = [
        'Merkur', 'Venus', 'Jorden', 'Mars', 'Jupiter',
        'Saturn', 'Uranus', 'Neptun', 'Pluto'
    ]
    planet_colors = [
        [.75, .75, .75], [1, 0.4, 0.6], [0, 0, 1], [1, 0, 0], [0, 1, 0],
        [0.7, 0, 0.7], [0.7, 0.7, 0], [0, 0.7, 0.7], [0.5, 0.5, 0.5]
    ]

    # Komet data
    halleys = [11, 17.8341442925537, 0.967142908462304,
               38.3842644764388, 2449400.5, 0.01308656479244564]

    # Vi laver et tomt solsystem
    S = Solar_system()
    # Vi tilføjer Solen
    S.add_sun(696340)
    # Vi tilføjer planeterne (og Pluto)
    S.add_planets(planet_names, planet_colors, planet_data)
    # Vi tilføjer Halleys komet
    S.add_comets(['Halleys komet'], [[0, 0, 0]], [halleys])
    # Vi laver passende skaleringer af solsystemet
    S.scale_radii(2 ** (-14))
    S.scale_distances(2)
    # Vi starter modellen til tiden 3000 e.Kr. Tallet 0.4 nedenfor afgør hvor
    # hurtigt planeterne bevæger sig i modellen.
    S.run(-3000, 0.4, 0.01)

File: Workshop 3 Solsystemet-20230210/solar_system.m
classdef solar_system<handle
    %Class for modeling the solar system
    
    properties
        sun
        planets
        comets
        t=-3000
        dt=10
        playing
        fig
        paused=0;
        axis_lim=[-40,40,-40,40];
        text
        show_text=1;
    end
    
    methods
        function obj = solar_system()
            %Constructer which initializes the figure window used to show
            %the model.
            obj.fig=figure(1);
            axis('equal')
            view(2);
            %hold off;
            axis equal;
            title('Solsystemet');
            axis(obj.axis_lim);
            set(gca,'visible','off')
            set(gca,'xtick',[])
            hold on;
            set(gcf,'WindowKeyPressFcn',@obj.buttonpush);
            obj.text=text(.8*obj.axis_lim(2),obj.axis_lim(4),{'q: afslut spillet',
    'mellemrum: pause/genoptag',
    'piletaster: skalér/zoom', 
    't: tekster til/fra',
    'punktum/komma: hastighed op/ned',
    ['dato: ',obj.time_to_date(obj.t)]}, 'verticalalign','top' );
        end
        
        
        
        function obj = add_planets(obj,names,colors,planet_data)
            %Method for adding planets to the solar system
            l=length(names);
            obj.planets=cell(1:l);
            %Sort the planets wrt radius
            [~,idx]=sort(planet_data(:,1));
            idx=flip(idx);
            names=names(idx);
            colors=colors(idx);
            planet_data=planet_data(idx,:);
            for i =1:l
                obj.planets{i}=Planet(names{i},colors{i},planet_data(i,:));
            end
        end
        function obj = add_comets(obj,names,colors,comet_data)
            %Method for adding comets to the solar system
            l=length(names);
            obj.comets=cell(1:l);
            %Sort the comets wrt radius
            sort(comet_data(:,1))
            [~,idx]=sort(comet_data(:,1));
            idx=flip(idx);
            names=names(idx);
            colors=colors(idx);
            for i =1:l
                obj.comets{i}=Comet(names{i},colors{i},comet_data(i,:));
            end
        end
        
        function obj=add_sun(obj,radius)
            %Method for adding a sun to the solar system
            obj.sun=Sun(radius);
        end
        
        function obj=update(obj,t)
            %Method for updating the solar system. This calculates the
            %positions of the objects in the solar system at the time t.
            obj.t=t;
            l=length(obj.planets);
            for i =1:l
                obj.planets{i}.update(t);
            end
            l=length(obj.comets);
            for i =1:l
                obj.comets{i}.update(t);
            end
            obj.sun.update();
        end
        function draw_system(obj)
            %Method for drawing the objects in the solar system at their
            %current coordinates.
            l=length(obj.planets);
            for i =1:l
                set(obj.planets{i}.ball, 'Visible', 'on')
                if obj.show_text && obj.axis_lim(1)<=obj.planets{i}.coordinates(1)&& obj.axis_lim(2)>=obj.planets{i}.coordinates(1)&&obj.axis_lim(3)<=obj.planets{i}.coordinates(2)&& obj.axis_lim(4)>=obj.planets{i}.coordinates(2)
                    set(obj.planets{i}.text,'Visible','on')
                else
                    set(obj.planets{i}.text,'Visible','off')
                end
            end
            l=length(obj.comets);
            for i =1:l
                set(obj.comets{i}.ball, 'Visible', 'on')
                if obj.show_text && obj.axis_lim(1)<=obj.comets{i}.coordinates(1)&& obj.axis_lim(2)>=obj.comets{i}.coordinates(1)&&obj.axis_lim(3)<=obj.comets{i}.coordinates(2)&& obj.axis_lim(4)>=obj.comets{i}.coordinates(2)
                    set(obj.comets{i}.text,'Visible','on')
                else
                    set(obj.comets{i}.text,'Visible','off')
                end
            end
            
            set(obj.sun.ball, 'Visible', 'on')
            if obj.show_text
                set(obj.sun.text,'Visible','on')
            else
                set(obj.sun.text,'Visible','off')
            end
            set(obj.text,'String', {'q: afslut spillet',
                'mellemrum: pause/genoptag',
                'piletaster: skalér/zoom', 
                't: tekster til/fra',
                'punktum/komma: hastighed op/ned',
                ['dato: ',obj.time_to_date(obj.t)]});
            set(obj.text,'Position',[0.9*obj.axis_lim(2),0.9*obj.axis_lim(4)]);
               
            drawnow
            axis('equal')
            axis(obj.axis_lim);
        end
        
        function obj=scale_distances(obj,scale)
             l=length(obj.planets);
            for i =1:l
                obj.planets{i}.a=obj.planets{i}.a*scale;
            end
            l=length(obj.comets);
            for i =1:l
                obj.comets{i}.a=obj.comets{i}.a*scale;
            end
        end
        
        function obj=scale_radii(obj,scale)
            l=length(obj.planets);
            for i =1:l
                obj.planets{i}.radius=obj.planets{i}.radius*scale;
            end
            
            l=length(obj.comets);
            for i =1:l
                obj.comets{i}.radius=obj.comets{i}.radius*scale;
            end
            obj.sun.radius=obj.sun.radius*scale;
        end
        
        
        function date=time_to_date(obj,t)
            %Method for converting the time of the system, measured in
            %years with 0 being 1BC to understandable date format.
            year=fix(t);
            fractional_part=t-year;
            day=abs(fix(365.25*fractional_part))+1;
            if year<1
                year=abs(year-1);
                prefix='BC';
            else 
                prefix='AD';
            end
            
            if 0<=day && day<=31
                month='Jan';
            elseif 31<day && day<=59
                month='Feb';
                day=day-31;
            elseif 59<day&& day<=90
                month='Mar';
                day=day-59;
            elseif 90<day && day<=120
                month='Apr';
                day=day-90;
            elseif 120<day&&day<=151
                month='May';
                day=day-120;
            elseif 151< day && day<=181
                month='Jun';
                day=day-151;
            elseif 181<day && day<=212
                month='Jul';
                day=day-181;
            elseif 212<day&&day<=243
                month='Aug';
                day=day-212;
            elseif 243<day&& day<=273
                month='Sep';
                day=day-243;
            elseif 273<day&& day<=304
                month='Oct';
                day=day-273;
            elseif 304<day &&day<= 334
                month='Nov';
                day=day-304;
            else 
                month='Dec';
                day=day-334;
            end
            if day>31
                day=31;
            end
            date=[num2str(day),'. ',month,' ',num2str(year),' ',prefix];
        end
        
        
        function run(obj,t,dt)
            %Method for running the model. This method updates the
            %positions for the planets and draws them for each time step
            obj.playing=1;
            obj.dt=dt;
            tic;
            while obj.playing==1 && t<=3000
                if obj.paused
                    obj.draw_system();
                    continue
                end
                obj.update(t);
                obj.draw_system();
                time_elapsed=toc;
                toc
                t=t+obj.dt*time_elapsed;
            end
            close(1)
        end
        
        function buttonpush(obj,src,ed)
            %Method for handling button pushes while the model is running.
            switch ed.Key
                case 'q'
                    obj.playing=0;
                case 'space'
                    if obj.paused
                        obj.paused=0;
                    else
                        obj.paused=1;
                    end
                case 'leftarrow'
                    obj.scale_radii(0.5);
                case 'rightarrow'
                    obj.scale_radii(2);
                case 'uparrow'
                    obj.axis_lim=obj.axis_lim*2;
                case 'downarrow'
                    obj.axis_lim=obj.axis_lim/2;
                case 'comma'
                    obj.dt=obj.dt/2;
                    for i =1:length(obj.planets)
                        obj.planets{i}.trace_length=obj.planets{i}.trace_length+20;
                    end
                case 'period'
                    obj.dt=obj.dt*2;
                    for i =1:length(obj.planets)
                        obj.planets{i}.trace_length=obj.planets{i}.trace_length-20;
                    end
                case 't'
                    if obj.show_text
                        obj.show_text=0;
                    else 
                        obj.show_text=1;
                    end
                    obj.show_hide_text();
            end
        end
    end
end

